fsw/cfe-core/src/sb/cfe_sbr.h

```diff
@@ -0,0 +1,236 @@
+    CFE_SB_RouteId_Atom_t RouteIdxTop;                             /**< \brief First unused entry in RoutingTbl */
+} cfe_sbr_t;
+
+/** \brief For each id callback function prototype */
+typedef void (*CFE_SBR_CallbackPtr_t)(CFE_SBR_RouteId_t RouteId, void *ArgPtr);
+
+/******************************************************************************
+ * Function prototypes
+ */
+
+/**
+ *  \brief Initialize software bus routing module
+ */
+void CFE_SBR_Init(void);
+
+/**
+ *  \brief Add a route for the given a message id
+ *
+ *  Called for the first subscription to a message ID, uses up one
+ *  element in the routing table.  Assumes check for existing
+ *  route was already performed or routes could leak
+ *
+ *  \param[in] MsgId Message ID of the route to add
+ *
+ *  \returns Route ID
+ */
+CFE_SBR_RouteId_t CFE_SBR_AddRoute(CFE_SB_MsgId_t MsgId);
+
+/**
+ *  \brief Obtain the route id given a message id
+ *
+ *  \param[in] MsgId Message ID of the route to get
+ *
+ *  \returns Route ID
+ */
+CFE_SBR_RouteId_t CFE_SBR_GetRouteId(CFE_SB_MsgId_t MsgId);
+
+/**
+ *  \brief Obtain the message id given a route id
+ *
+ *  \param[in] RouteId Route ID of the message id to get
+ *
+ *  \returns Message ID
+ */
+CFE_SB_MsgId_t CFE_SBR_GetMsgId(CFE_SBR_RouteId_t RouteId);
+
+/**
+ *  \brief Obtain the destination list head pointer given a route id
+ *
+ *  \param[in] RouteId Route ID
+ *
+ *  \returns Destination list head pointer for the given route id
+ */
+CFE_SB_DestinationD_t *CFE_SBR_GetDestListHeadPtr(CFE_SBR_RouteId_t RouteId);
+
+/**
+ * \brief Set the destination list head pointer for given route id
+ *
+ * \param[in] RouteId Route Id
+ * \param[in] DestPtr Destination list head pointer
+ */
+void CFE_SBR_SetDestListHeadPtr(CFE_SBR_RouteId_t RouteId, CFE_SB_DestinationD_t *DestPtr);
+
+/* TODO add description */
+void CFE_SBR_IncrementSequenceCounter(CFE_SBR_RouteId_t RouteId);
+uint32 CFE_SBR_GetSequenceCounter(CFE_SBR_RouteId_t RouteId);
+
+/**
+ * \brief Call the supplied callback function for all routes in MsgId order
+ *
+ * For certain routing implementations this is slower than CFE_SBR_ForEachRouteId,
+ * but guaranteed to be called in order of increasing MsgId raw values.  Only
+ * invokes callback if route exists.
+ *
+ * \param[in]  CallbackPtr Function to invoke for each matching ID
+ * \param[in]  ArgPtr      Opaque argument to pass to callback function
+ */
+void CFE_SBR_ForEachMsgId(CFE_SBR_CallbackPtr_t CallbackPtr, void *ArgPtr);
+
+/**
+ * \brief Call the supplied callback function for all routes
+ *
+ * For certain routing implementations this is faster than CFE_SBR_ForEachMsgId,
+ * but should only be used when MsgId order doesn't matter. Invokes callback
+ * for each route in the table.
+ *
+ * \param[in]  CallbackPtr Function to invoke for each matching ID
+ * \param[in]  ArgPtr      Opaque argument to pass to callback function
+ */
+void CFE_SBR_ForEachRouteId(CFE_SBR_CallbackPtr_t CallbackPtr, void *ArgPtr);
```

Comment:
These "for each" routines are somewhat similar to what is needed for #139 ... but in order to throttle we'll need some way to limit the number - i.e. only do a partial range - then resume where you left off on the next call(s).

Reply:
Added in fa8a216